### ISSUE TITLE
Document newsletter subscription from other views

### ISSUE THREAD

**Issue by dokterbob** [dokterbob/django-newsletter#240]
By including the `SubscribeRequestForm` in other views it is simple to allow users to subscribe from other pages, if you make sure the submit happens to the newsletter subscribe view. 

When wrapping this around an AJAX request, you can do it without having to display the redirect.

It would be really nice to have a word or two on this in the documentation as this seems a commonly requested feature.

Ref: #215

**Comment by bld2104** [comment]
I'm extremely interested in instructions/a tutorial for this!

**Comment by dokterbob** [comment]
You and others. If I tell you how, could you add it to the documentation? :) 
-- 
Verstuurd vanaf mijn Android apparaat met K-9 Mail. Excuseer mijn beknoptheid.

**Comment by bld2104** [comment]
@dokterbob yes yes yes! I'm very happy to do it. I have familiarity with AJAX too so I should be able to pick up what you are saying. I imagine we start by creating the custom form on the page we want, then the JQuery would look something like this?

```
$('#IDOfForm').on('submit', function(event){

 event.preventDefault();
    // document.getElementById("congrats").style.display = "block";
            $.ajax({
        url : "get-facebook-page-id/", // the endpoint
        type : "POST", // http method
        data : { 
            email-address : $('input[name=email]').val()
}
        dataType: 'json',
        // handle a successful response
        success : function(json) {
//redirect to one of the django-newsletter views?
        },

        // handle a non-successful response
        error : function(xhr,errmsg,err) {
          $('#error').html("<div class='alert-box alert radius' data-alert>Oops! We have encountered an error: "+errmsg+
                " <a href='#' class='close'>&times;</a></div>"); // add the error to the dom
            console.log(xhr.status + ": " + xhr.responseText); // provide a bit more info about the error to the console
        }
    });
});
```

**Comment by dokterbob** [comment]
Maybe first start with the non-AJAX version.

In short, you’d start by passing an instance of SubscribeForm (for a particular newsletter) to the context. Also reverse the SubscribeView for said newsletter and pass it in the context. Then just render the Django form as you normally would, passing in the subscribeview URL as action.

From thereon, the AJAX version should be trivial, as in it’s probably explained on StackOverflow. Essentially, you just fill in the form fields and click the submit, programmatically.

**Comment by bld2104** [comment]
Oh boy that's a bit tougher for me. I did some googling to try to come up with something though. It this close: 

```
class HomePageView(TemplateView):
    template_name = 'index.html'
    
    models      = Subscription
    queryset    = Subscription.objects.all()

    form_class  = SubscribeRequestForm

    def get_success_url(self):
        return reverse('newsletter_subscribe_request', kwargs={'slug': '<newsletter-name>')

    def get_context_data(self, **kwargs):
        context = super(SubscribeRequestView, self).get_context_data(**kwargs)
        context['form'] = self.get_form()
        return context
```

and then the form:

       ```
 <form enctype="multipart/form-data" method="post" action="/newsletter/<newsletter-name>/subscribe">
            {% csrf_token %}
            {{ form.as_p }}
            <p><input id="id_submit" name="submit" value="{% trans "Subscribe" %}" type="submit" /></p>
        </form>
```

**Comment by dokterbob** [comment]
Something like that, yeah. But it can be a little simpler, I think:

```python
class HomeView(TemplateView):
    ... <general stuff>
    def get_context_data(self, **kwargs):
        context = super(CustomerProfile, self).get_context_data(**kwargs)

        newsletter = Newsletter.objects.get(slug="bla")

        context['newsletter_form'] = NewsletterForm(newsletter=newsletter)

        return context
```

Then just use `{% reverse %}` for the action URL. Although, given that it's just an email and a name field, I suppose just hardcoding the form might work as well. As long as you make sure that you use the right action URL - so feedback on the forms (i.e. errors) are properly visible.

**Comment by dokterbob** [comment]
(By the way, this is loosehanded untested code, so you might have to tinker before it works...)

**Comment by bld2104** [comment]
@dokterbob So I am looking through GitHub and trying to understand your code. I'm mediocre with Django but very willing to persevere through this if you'll have patience with me! I am EXCELLENT at writing docs and will do a super thorough job though once I can follow what's happening. Here's where I'm at:

1) I'm using Django 2.0 and wondering if some of these statements might look different
2) I'm not sure how the code you sent checks if someone is a logged-in user or not (which based on the newsletter views on Github seems to be an important first step).
3) I can't figure out how your code is actually telling the homepage which form to render

Barb

**Comment by dokterbob** [comment]
Your contribution is greatly appreciated, there are many experienced coders who suck at documenting - and besides that Open Source is about doing things together, learning as well as creating.

> 	• I'm using Django 2.0 and wondering if some of these statements might look different
Not much experience with 2.0 yet, couldn’t tell, honestly.

> 	• I'm not sure how the code you sent checks if someone is a logged-in user or not (which based on the newsletter views on Github seems to be an important first step).
It doesn’t. I chose to leave it out in this instance. But it would look somewhat like this:

        if self.request.user:
            subscription, subscription_created = Subscription.objects.get_or_create(
                newsletter=newsletter, user=self.request.user
            )

            newsletterform = NewsletterForm(
                instance=subscription
            )
        else:
            newsletterform = NewsletterForm(
                instance=subscription, newsletter=newsletter
            )


> 	• I can't figure out how your code is actually telling the homepage which form to render

This you do in the template. The code just puts `newsletterform` in the context. Then something like `{{ newsletterform.as_p }}` renders it - but there are many ways to render forms.

Thanks for asking good questions! Already looking forward to reading the answers in the docs! :)

Mathijs

**Comment by bld2104** [comment]
Sorry for the delay getting back to you on this. Somehow, I broke my site while trying to play with this. It won't deploy to Heroku anymore. I've been spending every minute trying to get my site working again. I will keep experimenting with this once I get my site fixed.

**Comment by dokterbob** [comment]
No hurries, no worries. It comes when it comes.

**Comment by bld2104** [comment]
@dokterbob I'm having so much trouble with this and now a few other things. I think it's because I'm including it in my project as an egg? I am using Django 2.0 so when I tried pip install it didn't work and had to do it from the egg. I spent all this time customizing the models and templates and for a while it was picking up my customizations and now it is not (I didn't move anything). I think all this might be easier once you guys release the 2.0 version? Is that right? Also is that coming soon?

**Comment by dokterbob** [comment]
Hey Barbara,

Have you tried installing from GitHub instead of PyPI (as per alternative instructions in readme)?

Mathijs 
-- 
Verstuurd vanaf mijn Android apparaat met K-9 Mail. Excuseer mijn beknoptheid.

**Comment by bld2104** [comment]
Yes, this is what is in my requirements.txt file for Heroku:

`-e git+http://github.com/dokterbob/django-newsletter.git#egg=django-newsletter`

This could be because I had an issue with Heroku. But now I am trying to go back and re-do all the customizations I made (I had gotten rid of the "name" field for subscription for example). It seems to have mostly kept the template changes thankfully, but removed some of my styling. I can't remember how I overwrote these things the first time. 

To overwrite the forms to get rid of the name, do I need to create a separate forms file in a 'newsletter' directory? Same question for this current thread we have going - in which file do I put that view you wrote?

**Comment by dokterbob** [comment]
In a separate app named anything but newsletter! :) 
-- 
Verstuurd vanaf mijn Android apparaat met K-9 Mail. Excuseer mijn beknoptheid.

**Comment by bld2104** [comment]
Yeah thats what I did, I copied the NewsletterForm into my regular app's forms.py (which is what I thought I did the first time too) and it doesn't seem to be picking it up.

For the original question, I made a test view and right now the error I'm getting is that it can't find CustomerProfile. Did you mean User? Although I don't see User anywhere else in the views doc on Github.

**Comment by dokterbob** [comment]
CustomerProfile was a custom User in the particular project that code came from. 
-- 
Verstuurd vanaf mijn Android apparaat met K-9 Mail. Excuseer mijn beknoptheid.

**Comment by bld2104** [comment]
Okay so for the first issue (getting a sign-up form on another page) I replaced 'CustomerProfile' with 'User' and got this error:

`super(type, obj): obj must be an instance or subtype of type`

What should be going here?

Also, for the second idea, any idea why its not picking up my custom form? I checked and 'newseletter' is below 'myapp' in Installed_Settings in settings.py. This is what I put in myapp.forms.py (basically copied what's on GitHub but just removed the name field)

```
from newsletter.models import Subscription

class NewsletterForm(forms.ModelForm):
    """ This is the base class for all forms managing subscriptions. """

    class Meta:
        model = Subscription
        fields = ('email_field',)

    def __init__(self, *args, **kwargs):

        assert 'newsletter' in kwargs, 'No newsletter specified'

        newsletter = kwargs.pop('newsletter')

        if 'ip' in kwargs:
            ip = kwargs['ip']
            del kwargs['ip']
        else:
            ip = None

        super(NewsletterForm, self).__init__(*args, **kwargs)

        self.instance.newsletter = newsletter

        if ip:
            self.instance.ip = ip
```

It's not giving me any errors, it's just not working.

**Comment by bld2104** [comment]
@dokterbob *nudge* :)

**Comment by dokterbob** [comment]
Sorry, took a little holiday. Hope to find time this weekend! 
-- 
Verstuurd vanaf mijn Android apparaat met K-9 Mail. Excuseer mijn beknoptheid.

**Comment by bld2104** [comment]
No worries! And thank you! :)

**Comment by dokterbob** [comment]
> super(type, obj): obj must be an instance or subtype of type

This error is literally what it means. If you're calling `super()`, the second argument should be an instance of `type`. Example:
```python
class YoMamma(Parent):
    def is_how_fat(self):
        return super(YoMamma, self).is_how_fat()
```
In the above example, you could use `super(Parent, self)` as well, but not some other unrelated class.

As for the other problem (and this one as well), I can't help much if I can't see your code. Please commit it to some branch of a fork of django-newsletter and refer it here - preferably linking to the specific line numbers and posting the full stacktrace.

**Comment by bld2104** [comment]
Ahhh I am so sorry I am not this advanced. To be honest when you said "If I tell you the answer will you document it" I thought you meant you were literally going to give a complete answer and then I'd just write up what you did. I don't mind working through it this way still, but I don't know how to do most of what you just wrote. I just forked the django-newsletter repository to my account. Do I need to set this up as a whole new django project or you want me to just try to make the changes in GitHub to the django-newsletter files? 

And then I'm not sure how to link to the specific line numbers and "post the full stacktrace".

Barb

**Comment by bld2104** [comment]
Okay let's do the embedding a sign-up form issue first and then we can come back to the other one. For that one, I tried to understand your last comment, and put "EmbedNewsletterSubscribeView" into the super() instead of User. That error went away and now it's saying nothing found for slug='bla'. I totally understand why it's doing that, but when I tried putting in new things for slug='bla' nothing was working so I am not sure what goes there.

I put all of the code I created for it in this gist: https://gist.github.com/bld2104/68b92c646fef54113f7793d68880b453

The error message is here:
https://gist.github.com/bld2104/d65b8b7090f25911677513409544f2a8

If this is not enough, I can add you to my website's repository temporarily.

**Comment by bld2104** [comment]
I HAVE A SOLUTION!! I attended a local Python meetup group and they helped me get the answer! It was actually WAY more simple than what were were trying to do. You literally simply add:

```
  <form enctype="multipart/form-data" method="post" action="/newsletter/main-newsletter/subscribe/">
            {% csrf_token %}
<p style="color:white"><label for="id_email_field">E-mail:</label> <input type="email" name="email_field" required="" id="id_email_field"></p>
            <button class="btn btn-white" id="id_submit" name="submit" value="Subscribe" type="submit">Subscribe</button>
        </form>
```

To where you want the signup form to be. THAT'S IT! Nothing in views. Nothing in URLs. I will put this in the docs.

**Comment by bld2104** [comment]
I just submitted the request to include the instructions in the docs, but I messed up the code formatting and can't figure out how to fix it. I added ` to try to get the markup to recognize the code but it didn't sorry!

**Comment by dokterbob** [comment]
Great! 

Yeah restructured text is messy. For code blocks, you'd have a line with:: followed by a newline, then the code block indented by 4 spaces. In-line might work as well, but might require three backticks, I don't remember myself.

Note that you want to get the newsletter URL dynamically by using the reverse template tag. In your documentation, I would also leave any and all formatting out. 
-- 
Verstuurd vanaf mijn Android apparaat met K-9 Mail. Excuseer mijn beknoptheid.

**Comment by bld2104** [comment]
@dokterbob okay sorry about that, I wasn't sure how to edit my commit but again a group of local developers helped me figure that out (they said just submit a new commit, which I did [here](https://github.com/bld2104/django-newsletter/commit/73d522015973a344203348cf1bdf05d680e1e479)).

I added {% reverse %} as the action, but I did NOT test this on my site! I only tested it where you manually put in the newsletter name that you want the person to register for. I'm honestly a bit nervous about whether {% reverse %} will work or not, I'm not sure how it will know which newsletter without a view but maybe I don't understand the inner workings of django-newsletter well enough. 

I also fixed the formatting so that code blocks appear correctly, I realized there was a "preview" option so I was able to see what it would look like before submitting. Let me know if there is anything else I can do!

**Comment by dokterbob** [comment]
Hey, please *do* test it on your site. That's a lot easier to fix than giving other users wrong information.

You would use this view: https://github.com/dokterbob/django-newsletter/blob/master/newsletter/urls.py#L19

And then substitute the `newsletter_slug` parameter. You can look up how to use `{% reverse %}` in the excellent Django documentation.

Great you figured out how to update your pull request! Next time it'll be easier, I promise!

**Comment by bld2104** [comment]
@dokterbob okay I will try, but personally I'm going to continue manually putting in the newsletter name. I think that's a bit safer (I think that's how MailChimp does it too...).

Since I found a working solution (the one without reverse) and created documentation for it as promised, I need to now make this a back burner item I have a ton of other work in the next few weeks. I will try to get it working with reverse eventually because I'm appreciative for your help of course, but at least there is a working option for now for those who need something quickly.

**Comment by dokterbob** [comment]
@bld2104 Your choice. 

I guess the exact thing that demarcates good code/documentation is whether it just 'works for me' or can be used in a more general setting.

You mention you already have the documentation, could you at least fire a pull request for it? For example, by creating a subsection in the 'Usage' document? https://github.com/dokterbob/django-newsletter/blob/master/docs/usage.rst

Thanks so much!

**Comment by dokterbob** [comment]
Ah, sorry, I realise now that you did actually push code to your branch. Could you please file your very first pull request? (Takes 5 minutes.)

**Comment by bld2104** [comment]
@dokterbob okay I think I just submitted the pull request (I'm not great with GitHub yet, let me know if it did not go through)

**Comment by bld2104** [comment]
@dokterbob I did a brand new pull request for the Usage file. I had to do a new one anyway since the previous commits had issues (one had formatting issues, one had the %reverse% thing)

**Comment by dokterbob** [comment]
Thanks! Closed by #250.
